src/model/formats/foreflight-reader.ts

```diff
@@ -0,0 +1,118 @@
+import {
+  ForeFlightChecklist,
+  ForeFlightChecklistContainer,
+  ForeFlightChecklistGroup,
+  ForeFlightChecklistItem,
+  ForeFlightChecklistMetadata,
+  ForeFlightChecklistSubgroup,
+} from '../../../gen/ts/foreflight';
+import {
+  Checklist,
+  ChecklistFile,
+  ChecklistGroup,
+  ChecklistGroup_Category,
+  ChecklistItem,
+  ChecklistItem_Type,
+} from '../../../gen/ts/checklist';
+import { ForeFlightFormatError } from './foreflight-format';
+import { ForeFlightUtils } from './foreflight-utils';
+
+export class ForeFlightReader {
+  public static async read(file: File): Promise<ChecklistFile> {
+    const plaintext = await ForeFlightUtils.decrypt(await file.arrayBuffer());
+    const container: ForeFlightChecklistContainer = JSON.parse(plaintext);
+
+    if (container.type !== ForeFlightUtils.CONTAINER_TYPE) {
+      throw new ForeFlightFormatError(`unknown checklist type '${container.type}'`);
+    }
+
+    if (!container.payload) throw new ForeFlightFormatError('missing checklist payload');
+
+    if (container.payload.schemaVersion !== ForeFlightUtils.SCHEMA_VERSION) {
+      throw new ForeFlightFormatError(`unknown checklist schema version '${container.payload.schemaVersion}'`);
+    }
+
+    if (!container.payload.metadata) {
+      throw new ForeFlightFormatError('missing checklist metadata');
+    }
+
+    return {
+      groups: ForeFlightReader.checklistGroupsToEFIS(container.payload.groups),
+      metadata: {
+        name: ForeFlightReader.getChecklistName(container.payload.metadata),
+        description: container.payload.metadata.detail || '',
+        defaultGroupIndex: 0,
+        defaultChecklistIndex: 0,
+        makeAndModel: '',
+        aircraftInfo: container.payload.metadata.tailNumber || '',
+        manufacturerInfo: '',
+        copyrightInfo: '',
+      },
+    };
+  }
+
+  /**
+   * Unfortunately, all fields in ForeFlight are optional, so take first one filled out or default
+   */
+  private static getChecklistName(metadata: ForeFlightChecklistMetadata): string {
```

Comment:
Nitpicking: this is the FILE name, not a checklist name (probably just a nomenclature difference).

In other formats we set this to just the uploaded filename (minus the extension) - should we keep that consistent?

Reply:
> Anyway, I added https://github.com/rdamazio/efis-editor/commit/e2599f9255845c8cbc621d9c8f69a26299e5920a with what I think is the least bad solution

Thanks, that LGTM.